### PR TITLE
[2690] Add preview screens for file attachments

### PR DESCRIPTION
**Pull request by bychkovdmitry**
https://github.com/GetStream/stream-chat-android/issues/2690

### 🎯 Goal

Add previews for file attachments with common file types like PDFs, video, etc.

### 🛠 Implementation details

Added previews for attachments:
- Audio / Video attachments: a simple player based on the ExoPlayer library.
- Document attachments: a web view with Google Drive preview
- Other attachments with links: a simple `Intent.ACTION_VIEW` intent

### 🎨 UI Changes

https://user-images.githubusercontent.com/9600921/146168038-8e2d440b-d84b-457f-a081-b6ea6e5c992c.mp4

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [ ] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-team or #compose-chat-sdk-team) (required)
- [x] PR targets the `develop` branch

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### CODE REVIEW

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/content/FileAttachmentContent.kt [hunk 1]
```diff
@@ -71,10 +73,18 @@ public fun FileAttachmentContent(
...
+    val context = LocalContext.current
+
     Surface(
         modifier = Modifier
             .padding(2.dp)
-            .fillMaxWidth(),
+            .fillMaxWidth()
+            .clickable {
+                AttachmentPreviewHandler
+                    .defaultAttachmentHandlers(context)
```
Comment (filbabic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/content/FileAttachmentContent.kt:84]: We're creatign these handlers every time we click something.

Can we move this to ChatTheme?

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/filepreview/MediaAttachmentPreviewHandler.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ */
+public class MediaAttachmentPreviewHandler(private val context: Context) : AttachmentPreviewHandler {
+
+    override fun canHandle(attachment: Attachment): Boolean {
+        if (attachment.assetUrl.isNullOrEmpty()) return false
+
+        val type = attachment.type
+        val mimeType = attachment.mimeType
+
+        return type == ModelType.attach_audio ||
```
Comment (filbabic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/filepreview/MediaAttachmentPreviewHandler.kt:19]: Same here - return if it's null

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/filepreview/MediaAttachmentPreviewHandler.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            type == ModelType.attach_video ||
+            mimeType?.contains(MIME_TYPE_VIDEO) == true ||
+            mimeType?.contains(MIME_TYPE_AUDIO) == true ||
+            // For compatibility with other client SDKs
+            mimeType?.contains(MIME_SUBTYPE_MP4) == true ||
+            mimeType?.contains(MIME_SUBTYPE_QUICKTIME) == true
+    }
+
+    override fun handleAttachmentPreview(attachment: Attachment) {
+        context.startActivity(MediaPreviewActivity.getIntent(context, attachment.assetUrl!!))
```
Comment (filbabic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/filepreview/MediaAttachmentPreviewHandler.kt:29]: I'd rather use requireNotNull here or somehing if possible.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/filepreview/AttachmentPreviewHandler.kt [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+     * @return True if the current handler can provide a preview for this attachment.
+     */
+    public fun canHandle(attachment: Attachment): Boolean
+
+    /**
+     * Shows a preview for this attachment using the current preview handler.
+     *
+     * @param attachment The attachment we want to show the preview for.
+     */
+    public fun handleAttachmentPreview(attachment: Attachment)
```
Comment (filbabic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/filepreview/AttachmentPreviewHandler.kt:24]: Maybe pass the URL here instead? If we're only going to use that

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/filepreview/UrlAttachmentPreviewHandler.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    override fun handleAttachmentPreview(attachment: Attachment) {
+        val url = getAttachmentUrl(attachment)
+        context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
+    }
+
+    private fun getAttachmentUrl(attachment: Attachment): String? {
+        return when (attachment.type) {
+            ModelType.attach_image -> {
+                when {
+                    attachment.titleLink != null || attachment.ogUrl != null || attachment.assetUrl != null -> {
```
Comment (filbabic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/filepreview/UrlAttachmentPreviewHandler.kt:27]: Maybe do something like:

```
when {
titleLink != null -> titleLink
...
}
```

**Review by filbabic — COMMENTED** [review]
Left some comments, LMK if you feel we can include these improvements!

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/filepreview/AttachmentPreviewHandler.kt (hunk 4 above)
Comment (bychkovdmitry) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/filepreview/AttachmentPreviewHandler.kt:24]: Not so sure. In the UI components there is a logic which shows audio logo for audio files. I don't show it as it covers the play/pause icon, but still we might want to do something similar in future. Also, would be nice to show file names for video/audio files. I would keep this generic.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/MediaPreviewActivity.kt [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import android.os.Bundle
+import androidx.appcompat.app.AppCompatActivity
+import com.devbrackets.android.exomedia.listener.OnPreparedListener
+import com.devbrackets.android.exomedia.ui.widget.VideoView
+import io.getstream.chat.android.compose.R
+
+/**
+ * An Activity playing attachments such as audio and video.
+ */
+public class MediaPreviewActivity : AppCompatActivity(), OnPreparedListener {
```
[the remark this answers is not in the document]
Comment (adasiewiczr) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/MediaPreviewActivity.kt:15]: We can move it to the common module if you want to reuse the same XML based Activity 🙈 
Also, keep in mind we want to refactor this part and get ride of exomedia dependency

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/filepreview/AttachmentPreviewHandler.kt (hunk 4 above)
Comment (adasiewiczr) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/filepreview/AttachmentPreviewHandler.kt:24]: The new `MediaAttachmentActivity` contains a header with the attachment's title. You can check latest UI components app for reference

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/MediaPreviewActivity.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (bychkovdmitry) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/MediaPreviewActivity.kt]: I'm implementing a separate Activity with Compose components instead.

**Review by MarinTolic — APPROVED** [review]
### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [x] UI Changes correct (before & after images)
- [x] Bugs validated (bugfixes)
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

**Note:** There's a bit of a discrepancy between how Android API 31 and Android API 29 deal with certain files such as .jpg or.gif  where Android API 31 will open internal preview inside of the app where as Android API 29 will open external preview in Google Chrome

**Review by filbabic — APPROVED** [review]
As per @MarinTolic's comments - I checked the behavior and implementation here and it seems great!

We might want to revisit this in the future and improve the behavior in case we find bugs or something, but overall, great job and thanks for adding my suggestions! :]

**Review by filbabic — APPROVED** [review]

**Pull request merged**